### ISSUE TITLE
aws_kms_grant: Error "grant not found for key id" during terraform apply on refresh state

### ISSUE THREAD

**Issue by dshkyra** [terraform-providers/terraform-provider-aws#12555]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version
* Terraform v0.12.20
* Terraform AWS provider v2.54

### Affected Resource(s)

* aws_kms_grant

### Terraform Configuration Files

```hcl
resource "aws_kms_key" "test" {
  description = "Test key for grant not found issue"
}

resource "aws_iam_role" "test" {
  name = "test-kms-grant-not-found-role"

  assume_role_policy = <<EOF
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Action": "sts:AssumeRole",
      "Principal": {
        "Service": "lambda.amazonaws.com"
      },
      "Effect": "Allow",
      "Sid": ""
    }
  ]
}
EOF
}

resource "aws_kms_grant" "test" {
  name              = "test-kms-grant-not-found"
  key_id            = aws_kms_key.test.key_id
  grantee_principal = aws_iam_role.test.arn
  operations        = ["Encrypt", "Decrypt", "GenerateDataKey"]
}
```

### Debug Output
N/A

### Panic Output
N/A

### Expected Behavior
When a KMS grant is not found, it's treated as revoked. Terraform shows the change in the plan and recreates it during the apply.

### Actual Behavior
When a KMS grant was not found, `terraform apply` failed during refreshing state with an error:
```
aws_kms_key.test: Refreshing state... [id=REDACTED_KEY_ID]
aws_iam_role.test: Refreshing state... [id=test-kms-grant-not-found-role]
aws_kms_grant.test: Refreshing state... [id=REDACTED_KEY_ID:REDACTED_GRANT_ID]

Error: grant REDACTED_GRANT_ID not found for key id: REDACTED_KEY_ID
```

### Steps to Reproduce

1. `terraform init`
2. `terraform apply`
3. Revoke the grant created during apply, e.g. with AWS CLI `aws kms revoke-grant --key-id <KEY_ID> --grant-id <GRANT_ID>`
3. `terraform apply`

### Important Factoids
N/A

### References
N/A

**Comment by bflad** [comment]
In this case, an underlying function during read is returning a `resource.NotFoundError` which is not handled in the `Read` function before returning it as an error. Adding something like the following to the `Read` function should trigger resource recreation:

```go
	grant, err := findKmsGrantByIdWithRetry(conn, keyId, grantId)

	if isResourceNotFoundError(err) {
		log.Printf("[WARN] %s KMS grant id not found for key id %s, removing from state file", grantId, keyId)
		d.SetId("")
		return nil
	}

	if err != nil {
		return err
	}

	if grant == nil {
		log.Printf("[WARN] %s KMS grant id not found for key id %s, removing from state file", grantId, keyId)
		d.SetId("")
		return nil
	}
```

This can be tested by setting up acceptance testing similar to others with `_disappears` naming.

**Comment by dshkyra** [comment]
Thanks a lot for your comment @bflad!
I'm going to put a PR for that.

**Comment by dshkyra** [comment]
@bflad I've created a PR. Thanks again for your comment, especially about mentioning `_disappears` naming. It helped for me to find examples and write an integration test.

**Comment by bflad** [comment]
The fix for this has been merged and will release with version 2.56.0 of the Terraform AWS Provider, later this week. Thanks to @dshkyra for the fix. 👍
